draft-ietf-quic-transport.md

```diff
@@ -2206,9 +2212,17 @@ The endpoint MUST use unpredictable data in every PATH_CHALLENGE frame so that
 it can associate the peer's response with the corresponding PATH_CHALLENGE.
 
 An endpoint MUST expand datagrams that contain a PATH_CHALLENGE frame to at
-least the smallest allowed maximum datagram size of 1200 bytes.  Sending UDP
-datagrams of this size ensures that the network path from the endpoint to the
-peer can be used for QUIC; see {{datagram-size}}.
+least the smallest allowed maximum datagram size of 1200 bytes, unless the
+anti-amplification limit for the path does not permit sending a datagram of
+this size.  Sending UDP datagrams of this size ensures that the network path
+from the endpoint to the peer can be used for QUIC; see {{datagram-size}}.
+
+If an endpoint is unable to expand the datagram payload to 1200 bytes, no
+padding is necessary.  However, this means that the path MTU will not be
+validated.  The endpoint MUST perform a second path validation with a datagram
+of at least 1200 bytes once a PATH_RESPONSE is successfully received or when
+enough bytes have been received on the path that sending the larger datagram
+will not result in exceeding the anti-amplification limit.
```

Comment:
I think we can change this MUST to a "can", pointing out that the most likely reason for sending a small-sized PATH_CHALLENGE is due to receiving a non-probing packet on an unvalidated path due to unintentional migration. And that in such case, it is unlikely to have an alternative path.

I think that's what @ianswett pointed out.

Reply:
I might argue that the principle that we have is to reject paths with MTUs below 1200. But in case of NAT rebinding, the path has already switched. It's too late to reject that path.

But to be practical, I think we need to consider how endpoints should behave when MTU of the new path is below 1200. The two options at stake are:
* At the moment, the PR states that path validation MUST fail. But because the server might send smaller packets using the reminder of the 3x window, communication between the endpoints might slowly proceed.
* If we omit MTU discovery, communication between the endpoints is likely to slowly proceed in similar way.

IMO, both options are equally dissatisfatory, and based on that view, my preference goes to the simpler solution (of omitting MTU discovery).

By mandating MTU discovery, do we have a better way of bailing out?

Reply:
When we discussed this as part of the handshake, the principle we applied was a preference for failing fast.  That is, the sooner the problem is identified and the connection abandoned, the less likely it is that the connection persists in a half-broken state for an extended period.

Isn't a mandate for MTU probing a clear application of that principle?

(Nothing stops someone from trying to muddle on, but you need to be smart and effectively do a sub-1200 PMTUD of sorts to find your limits; and you have to rely on the peer doing the same.  That seems fragile.)

Reply:
Thank you for walking me through. So you are saying that an endpoint MUST always validate path MTU when responding to an unintentional migration (e.g., NAT rebinding), using a timer as recommended in [section 8.2.4](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-failed-path-validation). When the timer expires, and if there's no other active paths, the endpoint closes the connection.

As complicated as it is, I agree that there's benefit in closing early, and that the design is consistent.

The only remaining question I have is the error code to be used when closing the connection. Should we be using NO_ERROR, or do we need a new error code?

Reply:
I guess what distinguishes this from other cases is that there is an opportunity to signal an error.

However, in general we have avoided signaling this, because [losing a path doesn't mean losing the connection](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#section-8.2.4-6).  

I think that this is an error case where NO_ERROR almost makes sense.  Though I wouldn't be opposed to defining a NO_VIABLE_PATH error for this specific case.

Reply:
NO_VIABLE_PATH sounds good to me.

This specific case is not about losing one path. It is about losing a connection amid data transfer. That's going to have negative consequences, and I think it would be great if we could have insights regarding if that actually happens.

Depending on how often we would see this, we might start considering changing minimum MTU requirement to a smaller number, or remove the path validation requirement upon NAT rebinding.